cpp/CMakeLists.txt

```diff
@@ -14,19 +14,61 @@
 # limitations under the License.
 #=============================================================================
 
-cmake_minimum_required(VERSION 3.12...3.17 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
+
+# If `CMAKE_CUDA_ARCHITECTURES` is not defined, build for all supported architectures. If
+# `CMAKE_CUDA_ARCHITECTURES` is set to an empty string (""), build for only the current
+# architecture. If `CMAKE_CUDA_ARCHITECTURES` is specified by the user, use user setting.
+
+# This needs to be run before enabling the CUDA language due to the default initialization behavior
+# of `CMAKE_CUDA_ARCHITECTURES`, https://gitlab.kitware.com/cmake/cmake/-/issues/21302
+if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES OR CMAKE_CUDA_ARCHITECTURES STREQUAL "ALL")
+  set(CUSPATIAL_BUILD_FOR_ALL_ARCHS TRUE)
+elseif(CMAKE_CUDA_ARCHITECTURES STREQUAL "")
+  set(CUSPATIAL_BUILD_FOR_DETECTED_ARCHS TRUE)
+endif()
 
-project(CUDA_SPATIAL VERSION 0.19.0 LANGUAGES C CXX CUDA)
+project(CUSPATIAL VERSION 0.19.0 LANGUAGES C CXX)
 
-if(NOT CMAKE_CUDA_COMPILER)
-  message(SEND_ERROR "CMake cannot locate a CUDA compiler")
-endif()
+# Needed because GoogleBenchmark changes the state of FindThreads.cmake,
+# causing subsequent runs to have different values for the `Threads::Threads` target.
+# Setting this flag ensures `Threads::Threads` is the same value in first run and subsequent runs.
+set(THREADS_PREFER_PTHREAD_FLAG ON)
 
 ###################################################################################################
-# - build type ------------------------------------------------------------------------------------
+# - build options ---------------------------------------------------------------------------------
+
+option(USE_NVTX "Build with NVTX support" ON)
+option(BUILD_TESTS "Configure CMake to build tests" OFF)
+option(BUILD_BENCHMARKS "Configure CMake to build (google) benchmarks" OFF)
+option(JITIFY_USE_CACHE "Use a file cache for JIT compiled kernels" ON)
+option(CUDF_USE_ARROW_STATIC "Build and statically link Arrow libraries" OFF)
```

Comment:
I think we can remove these now since they're only used by libcudf:
```suggestion
```

Reply:
Removing `CUDF_USE_ARROW_STATIC` gives me this error using a local libcudf build dir for cuspatial:
```
CMake Error at /home/ptaylor/dev/rapids/cudf/cpp/cmake/thirdparty/CUDF_GetArrow.cmake:125 (find_and_configure_arrow):
  find_and_configure_arrow Function invoked with incorrect arguments for
  function named: find_and_configure_arrow
```

Reply:
We should probably add `CUDF_USE_ARROW_STATIC` to the `cudf-build-config.cmake.in` file... :/

Reply:
Done to both.